server/rest/src/main/java/org/infinispan/rest/cachemanager/RestCacheManager.java

```diff
@@ -220,15 +233,52 @@ public CompletionStage<Collection<CacheDistributionInfo>> cacheDistribution(Stri
          return CompletableFuture.completedStage(Collections.singletonList(local));
       }
 
-      Map<Address, CacheDistributionInfo> distributions = new ConcurrentHashMap<>();
+      return requestMembers(dm.getCacheTopology().getMembers(), ecm -> {
+         Cache<?, ?> c = SecurityActions.getCache(cacheName, ecm);
+         return CacheDistributionInfo.resolve(c.getAdvancedCache());
+      });
+   }
+
+   public CompletionStage<CompleteKeyDistribution> getKeyDistribution(String cacheName, Object key, RestRequest request) {
+      AdvancedCache<Object, ?> cache = getCache(cacheName, request.keyContentType(), request);
+      checkCachePermission(cache, request.getSubject(), AuthorizationPermission.MONITOR);
+      return SecurityActions.cacheContainsKeyAsync(cache, key)
+            .thenCompose(exists -> {
+               if (!exists)
+                  return CompletableFuture.completedFuture(Collections.emptyList());
+
+               DistributionManager dm = SecurityActions.getDistributionManager(cache);
+               if (dm == null) {
+                  KeyDistributionInfo local = KeyDistributionInfo.resolve(cache, true);
+                  return CompletableFuture.completedFuture(Collections.singletonList(local));
+               }
+
+               LocalizedCacheTopology topology = dm.getCacheTopology();
+               DistributionInfo distribution = topology.getDistribution(key);
+               return requestMembers(distribution.readOwners(), ecm -> {
```

Comment:
I am still not a fan of asking other nodes for their topology information. This means we could have different nodes saying conflicting information, as their topologies may not be exactly the same. I would say just use the local node's topology information. All we really need from the other node is its name, right?

Reply:
The request to the other members is only because of the name, but since we added the `primary` property, I've started using each node's localized topology to retrieve this value. I'll update to use only the local node for this.